### PR TITLE
feat(router): allow configuring app base href via token

### PR DESCRIPTION
**Pull request by btford**
Rather than relying on `<base href>`, you can bind a string to `appBaseHrefToken` to define the root URL for your app.

### CODE REVIEW

File: modules/angular2/src/mock/browser_location_mock.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import {proxy, SpyObject} from 'angular2/test_lib';
+import {IMPLEMENTS} from 'angular2/src/facade/lang';
+import {EventEmitter, ObservableWrapper} from 'angular2/src/facade/async';
+import {BrowserLocation} from 'angular2/src/router/browser_location';
+
+@proxy
+@IMPLEMENTS(BrowserLocation)
```
Comment (vicb) [inline modules/angular2/src/mock/browser_location_mock.ts]: would the `implements` keyword work ?

File: modules/angular2/src/mock/browser_location_mock.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+export class DummyBrowserLocation extends SpyObject {
+  baseHref;
+  internalPath;
+  _subject: EventEmitter;
+  constructor() {
+    super();
+    this.internalPath = '/';
+    this.baseHref = '/';
+    this._subject = new EventEmitter();
```
Comment (vicb) [inline modules/angular2/src/mock/browser_location_mock.ts]: move init values at declaration place

File: modules/angular2/src/mock/browser_location_mock.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  simulatePopState(url) {
```
Comment (vicb) [inline modules/angular2/src/mock/browser_location_mock.ts]: add return types for all methods and args

File: modules/angular2/src/mock/browser_location_mock.ts (hunk 1 above)
Comment (btford) [inline modules/angular2/src/mock/browser_location_mock.ts]: no because Dart

**Comment by btford** [comment]
I've addressed all the issues; assuming CI passes I'm gonna merge this.

Speak now or forever hold your peace :bride_with_veil:

**Pull request merged**